Review branch

CiaranO-C opened the pull request.

CiaranO-C committed:
Write basic calculator functions, write CSS and
HTML

CiaranO-C committed:
write function to convert user entry to postfix

CiaranO-C committed:
calculate function working, some styling changes,
addition of brackets button, removal of % btn

CiaranO-C committed:
rework infix to postfix logic, didn't reorder
characters properly.

some minor styling changes

CiaranO-C committed:
format code, add more styling

write event listeners to allow keyboard use

create backspace button in html, write JS code to add
functionality to btn

CiaranO-C committed:
remove commented out code

odinn123 commented on script.js:
```diff
@@ -0,0 +1,157 @@
+        const result = shuntingYard(userEntry.textContent);
+        userEntry.textContent = result;
+    };
+});
```
if
else if
else if

the event key wont be equally divide and enter during the same run through of the listener

odinn123 replied:
veeery minor detail but line 30- not entirely clear what youre attempting to do with the slice operation, store it in a variable which indicates its purpose and then assign it to text content

otherwise, this is resolved and generally looks a lot cleaner- if you wanted to, you could implement a switch case and have, for example: 

case 'enter'
case 'Escape'
case 'backspace
default
//iterate through buttons 

with its current structure i would reshuffle the checks: 

if 'enter'
else if 'Escape'
else if 'backspace
else
//iterate through buttons 

instead of iterating and checking all the buttons before the if. When the event corresponds to a button, you will need to iterate through them, however if they click escape you are currently checking lengthOfButtonArray+1 times

while in this case its a micro optimisation and not necessary at all, its better to have your heaviest operations run only when necessary as opposed to every single time 

also, quick tip: 

 buttonList.some(button => {
        if (key !== button.value) return false
            userEntry.push(key);
            (key === '/')
            ? display.textContent += '÷'
             : display.textContent += key;
            return true
        };
    });
    
    the some operation bails early when an element is found matching true on whatever condition you pass it, suppose we have an array of 100 elements, and the very first element is the one that meets the condition- 
    
    //this will iterate 100 times despite the first element being the desired one
    array.forEach((element)=>something)
    
    //this will iterate once 
    array.some((element)=>something)

   theres also array.every which bails out when an element returns false, good for checking if everything is valid in an array and throwing errors if not, or permission checking and restricting access

odinn123 commented on script.js:
```diff
@@ -0,0 +1,157 @@
+        userEntry.textContent += event.key;
+    };
+
+    if (event.key === '/') userEntry.textContent += '÷';
```
reduce complexity, why convert / to divide? dont even have the key on my keyboard, if someone were refactoring and/or extending this it would make the process unnecessarily frustrating

odinn123 commented on script.js:
```diff
@@ -0,0 +1,157 @@
+
+    if (button.target.id === 'backspace') userEntry.textContent = userEntry.textContent.slice(0, -1);
+});
+
```
if 
else if
else if
.....

odinn123 commented on script.js:
```diff
@@ -0,0 +1,157 @@
+const oper = {
+    '+': {
+        fn: add,
+        prec: 1,
```
precendence

odinn123 commented on script.js:
```diff
@@ -0,0 +1,157 @@
+                stack.push(expr[i]);
+            } else if (expr[i] !== ')') {
+                while (stack[stack.length - 1] !== '(' && stack.length && oper[expr[i]].prec <= oper[stack[stack.length - 1]].prec) {
+                    output.push(stack.pop());
```
split it out onto multiple lines and store the value for the next index , its horrible to read: 

oper[stack[stack.length-1].prec

becomes 

const top =stack.length-1
const next= stack[top]
oper[next]

odinn123 commented on script.js:
```diff
@@ -0,0 +1,157 @@
+    if (stack[0].includes('-') && stack[0][0] !== '-') {
+        return shuntingYard(stack[0]);
+    } else {
+        let expression = stack.toString();
```
const

CiaranO-C committed:
separate display from logic

improve keydown and click event listener conditions

CiaranO-C committed:
reduce complexity of convertExpression function

write new function that evaluates converted expr

added functionality for escape and backspace keys

odinn123 commented on script.js:
```diff
@@ -107,68 +110,75 @@ const oper = {
+                stack.push(char);
+            } else if (oper.hasOwnProperty(char)) {
+                while (stack.length && stack[top] !== '(' && oper[char].precedence <= oper[stack[top]].precedence) {
                     output.push(stack.pop());
```
much cleaner, remember scope though- each iteration the 'top' variable is an entirely new box, unless you are reassigning it later (cant see the later code on this diff) it can be const

odinn123 commented on script.js:
```diff
@@ -107,68 +110,75 @@ const oper = {
-                while (stack[stack.length - 1] !== '(') {
+                while (stack[top] !== '(') {
                     output.push(stack.pop());
+                    top = stack.length - 1;
```
ignore my previous comment, just saw this, valid usage of let

odinn123 commented on script.js:
```diff
@@ -0,0 +1,184 @@
+    if (button.classList.contains('oper')) {
+        button.style.color = 'orange'
+    };
+};
```
style pojint, dont have to follow it 

if(!button.classList.contains...)return 

invert your conditions and return early where possible, means if you do need additional conditions after that point, they wont be nested and easier to read

odinn123 commented on script.js:
```diff
@@ -0,0 +1,184 @@
+                accum = '';
+            } else {
+                output.push(char); //pushes single digit numbers
+            };
```
here is a good example of my earlier mention of returning early, the 'else' is the only thing that takes place after the last condition, so if you do the equivalent of an early return in the last block, you don't need it. 

As you're in a loop, the equivalent is either break, or continue

break, breaks out the whole loop, you dont want this to happen here 

continue, continues to the next iteration of the loop, skipping all the code that follows it for this current iteration- this is the one you want

odinn123 commented on script.js:
```diff
@@ -0,0 +1,184 @@
+        return result;
+    };
+};
+
```
just leaving a note here, generally much cleaner, good job

CiaranO-C committed:
improve keydown logic, changed nodelist to array,
changed .forEach loop to .some, avoids iterating through all buttons
unnecessarily
